homeassistant/components/hunterdouglas_powerview/cover.py

```diff
@@ -341,7 +411,189 @@ class PowerViewShade(PowerViewShadeBase):
     def _async_process_updated_position_data(self, position_data):
         """Process position data."""
         if ATTR_POSITION1 in position_data:
-            self._current_hd_cover_position = int(position_data[ATTR_POSITION1])
+            self._current_hd_cover_primary = int(position_data[ATTR_POSITION1])
+
+
+class PowerViewShadeTDBU(PowerViewShade):
+    """Representation of a PowerView shade with top/down bottom/up capabilities."""
+
+    @property
+    def current_cover_position_secondary(self):
+        """Return the current position of cover."""
+        return hd_position_to_hass(self._current_hd_cover_secondary)
+
+    async def async_open_cover(self, **kwargs):
+        """Close the cover."""
+        current_hass_pos1 = self.current_cover_position
+        current_hass_pos2 = self.current_cover_position_secondary
+        steps_to_move = 100 - (current_hass_pos1 + current_hass_pos2)
+        self._async_schedule_update_for_transition(steps_to_move)
+        self._async_update_from_command(await self._shade.open())
+
+    async def async_close_cover(self, **kwargs):
+        """Close the cover."""
+        current_hass_pos1 = self.current_cover_position
+        current_hass_pos2 = self.current_cover_position_secondary
+        steps_to_move = current_hass_pos1 + current_hass_pos2
+        self._async_schedule_update_for_transition(steps_to_move)
+        self._async_update_from_command(await self._shade.close())
+
+    @callback
+    def _async_process_updated_position_data(self, position_data):
+        """Process position data."""
+        if ATTR_POSITION1 in position_data:
+            if int(position_data[ATTR_POSKIND1]) == POSKIND_PRIMARY:
+                self._current_hd_cover_primary = int(position_data[ATTR_POSITION1])
+        if ATTR_POSITION2 in position_data:
+            if int(position_data[ATTR_POSKIND2]) == POSKIND_SECONDARY:
+                self._current_hd_cover_secondary = int(position_data[ATTR_POSITION2])
+
+
+class PowerViewShadeTDBUBottom(PowerViewShadeTDBU):
+    """Representation of a top down bottom up powerview shade."""
+
+    def __init__(self, coordinator, device_info, room_name, shade, name):
+        """Initialize the shade."""
+        super().__init__(coordinator, device_info, room_name, shade, name)
+        self._unique_id = f"{self._shade.id}-bottom"
```

Comment:
You'll have to adjust the base class since its being overridden there as well

```
diff --git a/homeassistant/components/hunterdouglas_powerview/entity.py b/homeassistant/components/hunterdouglas_powerview/entity.py
index 70ecb5d79c..4528f8f37e 100644
--- a/homeassistant/components/hunterdouglas_powerview/entity.py
+++ b/homeassistant/components/hunterdouglas_powerview/entity.py
@@ -29,14 +29,9 @@ class HDEntity(CoordinatorEntity):
         """Initialize the entity."""
         super().__init__(coordinator)
         self._room_name = room_name
-        self._unique_id = unique_id
+        self._attr_unique_id = unique_id
         self._device_info = device_info
 
-    @property
-    def unique_id(self):
-        """Return the unique id."""
-        return self._unique_id
-
     @property
     def device_info(self) -> DeviceInfo:
         """Return the device_info of the device."""
```

Then these can use `self._attr_unique_id` instead
```
cover.py:        self._unique_id = f"{self._shade.id}-bottom"
cover.py:        return f"{self._unique_id}_bottom"
cover.py:        self._unique_id = f"{self._shade.id}-top"
cover.py:        return f"{self._unique_id}_top"
sensor.py:        return f"{self._unique_id}_charge"
```

Reply:
thanks - didnt think about the entity.py file - when you mentioned BAse i thought you meant the new Base Class.

I have removed `def unique_id(self):` but in doing so also needed to add the below to sensor.py
`
    def __init__(self, coordinator, device_info, room_name, shade, name):
        """Initialize the shade."""
        super().__init__(coordinator, device_info, room_name, shade, name)
        self._attr_unique_id = f"{self._attr_unique_id}_charge"`